### DISCUSSION TITLE
How to subscribe to onFailure events when using the circuit breaker strategy in the ResilienceCommand class

### DISCUSSION THREAD

**Discussion by fhosen** [SocketSomeone/nestjs-resilience#328] [Q&A] [answered]
Hello! First of all thanks for you work 🙇‍♂️ 

With my team we are considering to use this tool in our API, so we are performing some tests.

We were playing around with the decorator and the ResilienceCommand. And we started thinking on how could we eventually export stats to gain observability, similar to what [this lib](https://github.com/nodeshift/opossum-prometheus) does for Opossum.

So we wanted to subscribe to some events, and we could do it for the success event, but not for the failure or short circuit.
We tried to do so in the constructor and also overriding some of the methods. I'll share with you the code snippet, but really I wanted to ask you if there is a right way to do this:

```typescript
import { Injectable } from '@nestjs/common';
import { ResilienceCommand, ResilienceFactory, ResilienceEventType } from 'nestjs-resilience';
import { ClientsService } from './clients.service';

@Injectable()
export class GetClientsService extends ResilienceCommand {
  constructor(
    private readonly factory: ResilienceFactory,
    private readonly clientsService: ClientsService
  ) {
    const circuitBreakerStrategy = factory.createCircuitBreakerStrategy({
      requestVolumeThreshold: 1,
      sleepWindowInMilliseconds: 5000,
      errorThresholdPercentage: 50
    });
    super([circuitBreakerStrategy]);
    this.eventBus.on(ResilienceEventType.ShortCircuit, () => console.log('GetClientsService EventBus:ShortCircuit'))
    this.eventBus.on(ResilienceEventType.Success, () => console.log('GetClientsService EventBus:Success'))
    this.eventBus.on(ResilienceEventType.Failure, () => console.log('GetClientsService EventBus:Failure'))
  }

  async run(id: number) {
    console.log('GetClientsService run')
    return this.clientsService.findOneById(id);
  }
  
  override onFailure(error: any) {
    console.log('GetClientsService onFailure')
    return super.onFailure(error);
  }

  override onSuccess() {
    console.log('GetClientsService onSuccess')
    return super.onSuccess();
  }
}

```

**Comment by SocketSomeone** [comment, marked as answer]
@fhosen Hi there!

Thank you for question! With your question I found some issues with events and fixed them. (https://github.com/SocketSomeone/nestjs-resilience/releases/tag/2.1.1)
Now it`s works better:
1. Failure and ShortCircuit events will be fired
2. EventBus using singleton and you can access to them with `ResilienceEventBus.getInstance()` (https://github.com/SocketSomeone/nestjs-resilience/blob/master/test/commands/resilience.command.spec.ts#L53)

Btw, if you wanna create metrics storage by commands, feel free to open PR!

**Comment by fhosen** [comment]
Hey @SocketSomeone, thanks for the quick answer and fix; it's working like a charm now 😄 

> Btw, if you wanna create metrics storage by commands, feel free to open PR!

If we end up implementing something like that, for sure we'll try to contribute to the project!
